Issue #1325: (lib-fetch-httpurlconnection): A concept-fetch implementation using HttpURLConnection.

pocmo opened the pull request:
~~Depends on PR #1330.~~

~~**Let's review and land PR #1330 before this one (after rebasing this PR will just have one commit)**.~~

Ready for review!

pocmo commented:
It looks like one of the timeout tests is running indefinitely on taskcluster :)

csadilek commented on components/lib/fetch-httpurlconnection/src/main/java/mozilla/components/lib/fetch/httpurlconnection/HttpURLConnectionClient.kt:
```diff
@@ -0,0 +1,154 @@
+
+    var index = 0
+
+    while (connection.getHeaderField(index) != null) {
```
Would it be easier to use `connection.headerFields` and iterate over it, rather than using the index?

pocmo commented:
Addressed the review comments. Still need to figure out why one of the tests is hanging. It doesn't locally and the same test passes for the OkHttp implementation.

pocmo commented:
I had to remove the connect timeout test. It was behaving as I expected (it looks like my assumption that this scenario would trigger a connect timeout is wrong) and it was super flaky even locally.

Only an unrelated component failed (jcenter yay) so I'm going ahead and merge.
